Add customizable arrow colors to FlashcardArray

mgiannopoulos24 opened the pull request:
### Description

This PR introduces a new feature that allows developers to customize the colors of the navigation arrows in the `FlashcardArray` component. This provides greater flexibility for styling and theming the component to match different application designs.

### Related issues
Closes #17.

### Changes
1. Updated `useFlashcardArray` Hook:
    - The UseFlashcardArrayProps interface now accepts two new optional properties:
        - arrowColor (string, default: `#1c1b1e`)
        - disabledArrowColor (string, default: `#a9a9a9`)
    - The hook now returns these color values, making them accessible to the component layer.
2. Modified `FlashcardArray` Component:
    - The component's navigation buttons now dynamically apply the `arrowColor` or `disabledArrowColor` to the fill style of the SVGs based on the button's disabled state.
3. Added Storybook Demonstration:
    - A new story, `CustomArrowColors`, has been added to FlashcardArray.stories.tsx. 4. Updated Unit Tests:
    - Tests for the `useFlashcardArray` hook have been updated to verify that the new color props are initialized correctly, both with default and custom values.
    - Tests for the `FlashcardArray` component have been enhanced to include a test case that asserts the correct colors are applied to the SVG arrows based on their enabled/disabled state. 

### Showcase
<p align="center">
<img src="https://imgur.com/U4K8AW9.png">
</p>

As you can see on the left is the colored arrow and on the right is the disabled colored arrow.

---

Let me know your thoughts on this PR.

mgiannopoulos24 committed:
Add customizable arrow colors FlashcardArray

ABSanthosh commented on package/src/components/FlashcardArray/index.tsx:
This implementation could be simpler. In the button's `disabled` attribute, we can check if `localFlipArrayHook.cardsInDisplay[0]===-1` for prev button and `localFlipArrayHook.cardsInDisplay[1]==-1` for next button. The way it is set up, if cycle is off, the `cardsInDisplay` array starts with `[-1, 0, 1]` and ends with `[5, 6, -1]`. If the cycle is on, its `[6, 0, 1]` and `[5, 6, 0]`.

ABSanthosh replied:
This could be put back in the `useFlashcardArray` component so developers who use the hook can read from this instead of doing the check manually. Can we add `canGoPrev(): boolean` and `canGoNext(): boolean` to the hook with the logic I mentioned above and return them from the hook?

ABSanthosh commented:
Hi @mgiannopoulos24, thanks for your PR.

In case the user wants to customize the arrow colors, they could use the following scss or css selector or make a custom button with the desired color.
```scss
.flashcard-array__controls {
  & > button > svg > path {
    fill: var(--foreground) !important;
  }
}
``` 
I'm not sure if it is necessary to add more props when there are multiple ways to achieve that. An alternative, non-js way to let user pick a color would be to use css variables like how I have for the flashcard front and back cards' background:
```scss
.flashcard-wrapper {
  --border-radius: 16px;
  --back-bg: #ffffff;
  --front-bg: #ffffff;
  --box-shadow: 0 0 2.5rem 0 rgba(0, 0, 0, 0.16);
...
}
```
I'm suggesting a non-js way because there would be very few instances where a dev would want to change arrow color dynamically. 

I like the disabled state for the svg arrow. We just need to generalize it for others to use as well.

mgiannopoulos24 committed:
Hook update and simplified code

mgiannopoulos24 commented:
I've updated the useFlashcardArray hook to make handling navigation easier.

I added two new boolean properties, `canGoPrev` and `canGoNext`, to the hook. These properties figure out if it's possible to navigate backward or forward.

I'm not very sure about the arrow colors. As a developer I would prefer to pass them as props and not to change a CSS file, but if you are not okay with that I can remove it completely.

ABSanthosh commented:
I was looking more into DX like this
```ts
<FlashcardArray
  style={
    {
      '--prev-arrow-color': '#1c1bee',
      '--next-arrow-color': '#1c1bee',
      '--disabled-arrow-color': '#a9a9a9',
    } as CSSProperties
  }
  flipArrayHook={flipArrayHook}
  deck={deck.cards.map((card) => ({
    id: card.id,
    front: card.front,
    back: card.back,
  }))}
/>
```

`/react-quizlet-flashcard/package/src/components/FlashcardArray/style.scss`
```scss
.flashcard-array-wrapper {
  --prev-arrow-color: #4614a9;
  --next-arrow-color: #4614a9;
  --disabled-arrow-color: #a9a9a9;
  ...
}
...
&__controls {
  ...
  button {
    svg {
      path {
        &[data-name='Left'] {
          fill: var(--prev-arrow-color);
        }

        &[data-name='Right'] {
          fill: var(--next-arrow-color);
        }
      }
    }
    &:disabled {
      cursor: not-allowed;
      svg path {
        fill: var(--disabled-arrow-color);
      }
    }
  }
}
```

You also have to remove the color from inline styles on `path` element in FlashcardArray component.

Since this style is passed to the `flashcard-array-wrapper` div, it would be scoped to that component. I just don't think using js to set the styles is necessary when it can be achieved directly through css. What do you think?

mgiannopoulos24 committed:
Changed to request

mgiannopoulos24 commented:
After thinking it for a while you are indeed correct. We don't need to overcomplicate it. 

The only thing I wasn't able to understand is do you want me to change something to `src/components/FlashcardArray/style.scss`?

mgiannopoulos24 committed:
Fix SCSS file

mgiannopoulos24 commented:
I've read it again and I fixed it.

ABSanthosh committed:
feat: added docs for custom arrow style

ABSanthosh committed:
chore: add class for flashcard-array-wrapper and custom arrow css variables

ABSanthosh committed:
chore: build and ver bump

ABSanthosh commented:
I made some changes. We don't need the `!important`s so I removed them, added docs for the arrow color change and version bumped. Not sure if I'm going to publish this right away. #23 and #21 are relatively easier to add too so either once someone makes a PR or when I work on it, I can publish it at one go.

mgiannopoulos24 commented:
Okay I will try to implement both #23 and #21. Are you okay with that?

ABSanthosh commented:
Sure. For #21, I can't reproduce the animation overflow anymore. I was seeing that issue during dev but for the docs, I removed the width and height and set both to 100% and added max width and max height so it always fills the size but also don't overflow. We also need to consider what the developers might expect when the card size is bigger than the view port. The card content may not be responsive so it may break that if we shrink the card and not let the developer control the way it handles responsiveness.

I'm merging this PR to master now.
